runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,298 @@
+        return new CorfuRecordId(0, id);
+    }
+
+    /**
+     * CorfuQueueRecord encapsulates each entry enqueued into CorfuQueue with its unique ID.
+     * It is a read-only type returned by the entryList() method.
+     * The ID returned here can be used for both point get()s as well as remove() operations
+     * on this Queue.
+     *
+     * @param <E>
+     */
+    public class CorfuQueueRecord<E> implements Comparable<CorfuQueueRecord<E>> {
+        /**
+         * This ID represents the entry and its order in the Queue.
+         * This implies that it is unique and comparable with other IDs
+         * returned from CorfuQueue methods with respect to its enqueue order.
+         * However it cannot be compared with ID returned from the enqueue method for ordering
+         * because if this method is wrapped in a transaction, the order is established only later.
+         */
+        @Getter
+        final private CorfuRecordId recordId;
+
+        @Getter
+        final private E entry;
+
+        public String toString() {
+            return String.format("%s=>%s", recordId, entry);
+        }
+
+        CorfuQueueRecord(long ordering, long entryId, E entry) {
+            this.recordId = new CorfuRecordId(ordering, entryId);
+            this.entry = entry;
+        }
+
+        @Override
+        public int compareTo(CorfuQueueRecord<E> o) {
+            return this.recordId.compareTo(o.getRecordId());
+        }
+    }
+
+    /**
+     * We need to encode 2 pieces of information into an 8 byte long to represent ordering.
+     * 1. We use 40 bits for CorfuQueue's snapshot version information.
+     * 2. We use 24 bits for the index within that CorfuQueue's snapshot.
+     */
+    final private static int MAX_BITS_FOR_INDEX = 24;
+
+    /**
+     * Returns a List of CorfuQueueRecords sorted by the order in which the enqueue materialized.
+     * This is the primary method of consumption of entries enqueued into CorfuQueue.
+     *
+     * To re-constitute the commit order in the presence of transactions the CorfuRecords returned
+     * here have their UUID fields composed of two parts:
+     *     +----------------------------------------------------------------------------+
+     *     | Stream snapshot |Index in snapshot|     ID of the entry in the map        |
+     *     +----------------------------------------------------------------------------+
+     *     <----5 bytes------><----3 bytes-----><------- 8 bytes ----------------------->
+     *
+     * <p>This function currently does not return a view like the java.util implementation,
+     * and changes to the entryList will *not* be reflected in the map. </p>
+     *
+     * @param maxEntries - Limit the number of entries returned from start of the queue
+     * @throws IllegalArgumentException if maxEntries is negative.
+     * @return List of Entries sorted by their enqueue order
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntries) {
```

Comment:
Is it possible to have List<CorfuQueueRecord<? super E>>?

Reply:
great suggestion again. thank you